### ISSUE TITLE
Error: Unable to synchronize and save tokens on nodes: mapping, Corosync, Error, Nodes:. Are they authorized?

### ISSUE THREAD

**Issue by rpolasek** [ClusterLabs/pcs#247]
**os:**
  CentOS Linux release 7.8.2003 (Core)

**rpm:**
  pcs-0.9.168-4.el7.centos.x86_64
  corosync-2.4.5-4.el7.x86_64
  pacemaker-1.1.21-4.el7.x86_64

**error messags:**
c7-velleros-sde2a: Unable to authenticate to c7-velleros-sde2a - (HTTP error: 401), try running 'pcs cluster auth'
c7-velleros-sde2b: Unable to authenticate to c7-velleros-sde2b - (HTTP error: 401), try running 'pcs cluster auth'

**cat /etc/hosts**
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6
192.168.15.34 c7-velleros-sde2a
192.168.15.35 c7-velleros-sde2b

**cat /etc/corosync/corosync.conf**
```
totem {
  version: 2
  cluster_name: sdecmsp
  interface {
    ringnumber: 0
    bindnetaddr: 192.168.15.0
    mcastport: 5405
  }
  secauth: off
  transport: udpu
}

logging {
  fileline: off
  to_logfile: yes
  to_syslog: yes
  to_stderr: no
  logfile: /var/log/cluster/corosync.log
  debug: off
  timestamp: off
  logger_subsystem {
    subsys: QUORUM
    debug: off
  }
}

nodelist {
        node {
                node_id: 1
                ring0_addr: c7-velleros-sde2a
        }
        node {
                node_id: 2
                ring0_addr: c7-velleros-sde2b
        }

}

quorum {
  provider: corosync_votequorum
  two_node: 1
}
```

**pcs cluster auth c7-velleros-sde2a c7-velleros-sde2b -uhacluster -p'*****' --debug**
```
Running: /usr/bin/ruby -I/usr/lib/pcsd/ /usr/lib/pcsd/pcsd-cli.rb auth
Environment:
  GEM_HOME=/usr/lib/pcsd/vendor/bundle/ruby
  HISTCONTROL=ignoredups
  HISTSIZE=1000
  HISTTIMEFORMAT=%F %T 
  HOME=/root
  HOSTNAME=c7-velleros-sde2a
  LANG=en_US.UTF-8
  LC_ALL=C
  LC_COLLATE=C
  LESSOPEN=||/usr/bin/lesspipe.sh %s
  LOGNAME=root
  LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=01;05;37;41:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.axv=01;35:*.anx=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=01;36:*.au=01;36:*.flac=01;36:*.mid=01;36:*.midi=01;36:*.mka=01;36:*.mp3=01;36:*.mpc=01;36:*.ogg=01;36:*.ra=01;36:*.wav=01;36:*.axa=01;36:*.oga=01;36:*.spx=01;36:*.xspf=01;36:
  MAIL=/var/spool/mail/root
  PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin:/root/bin
  PCSD_DEBUG=true
  PCSD_NETWORK_TIMEOUT=60
  PWD=/root
  SHELL=/bin/bash
  SHLVL=2
  SSH_AUTH_SOCK=/tmp/ssh-bm4D1omowc/agent.16899
  SSH_CLIENT=10.0.5.4 44782 22
  SSH_CONNECTION=10.0.5.4 44782 192.168.15.34 22
  SSH_TTY=/dev/pts/0
  STY=16937.pts-0.c7-velleros-sde2a
  TERM=screen
  TERMCAP=SC|screen|VT 100/ANSI X3.64 virtual terminal:\
        :DO=\E[%dB:LE=\E[%dD:RI=\E[%dC:UP=\E[%dA:bs:bt=\E[Z:\
        :cd=\E[J:ce=\E[K:cl=\E[H\E[J:cm=\E[%i%d;%dH:ct=\E[3g:\
        :do=^J:nd=\E[C:pt:rc=\E8:rs=\Ec:sc=\E7:st=\EH:up=\EM:\
        :le=^H:bl=^G:cr=^M:it#8:ho=\E[H:nw=\EE:ta=^I:is=\E)0:\
        :li#84:co#148:am:xn:xv:LP:sr=\EM:al=\E[L:AL=\E[%dL:\
        :cs=\E[%i%d;%dr:dl=\E[M:DL=\E[%dM:dc=\E[P:DC=\E[%dP:\
        :im=\E[4h:ei=\E[4l:mi:IC=\E[%d@:ks=\E[?1h\E=:\
        :ke=\E[?1l\E>:vi=\E[?25l:ve=\E[34h\E[?25h:vs=\E[34l:\
        :ti=\E[?1049h:te=\E[?1049l:us=\E[4m:ue=\E[24m:so=\E[3m:\
        :se=\E[23m:mb=\E[5m:md=\E[1m:mr=\E[7m:me=\E[m:ms:\
        :Co#8:pa#64:AF=\E[3%dm:AB=\E[4%dm:op=\E[39;49m:AX:\
        :vb=\Eg:G0:as=\E(0:ae=\E(B:\
        :ac=\140\140aaffggjjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~..--++,,hhII00:\
        :po=\E[5i:pf=\E[4i:Km=\E[M:k0=\E[10~:k1=\EOP:k2=\EOQ:\
        :k3=\EOR:k4=\EOS:k5=\E[15~:k6=\E[17~:k7=\E[18~:\
        :k8=\E[19~:k9=\E[20~:k;=\E[21~:F1=\E[23~:F2=\E[24~:\
        :F3=\E[1;2P:F4=\E[1;2Q:F5=\E[1;2R:F6=\E[1;2S:\
        :F7=\E[15;2~:F8=\E[17;2~:F9=\E[18;2~:FA=\E[19;2~:kb=:\
        :K2=\EOE:kB=\E[Z:kF=\E[1;2B:kR=\E[1;2A:*4=\E[3;2~:\
        :*7=\E[1;2F:#2=\E[1;2H:#3=\E[2;2~:#4=\E[1;2D:%c=\E[6;2~:\
        :%e=\E[5;2~:%i=\E[1;2C:kh=\E[1~:@1=\E[1~:kH=\E[4~:\
        :@7=\E[4~:kN=\E[6~:kP=\E[5~:kI=\E[2~:kD=\E[3~:ku=\EOA:\
        :kd=\EOB:kr=\EOC:kl=\EOD:km:
  USER=root
  WINDOW=2
  XDG_RUNTIME_DIR=/run/user/0
  XDG_SESSION_ID=34
  _=/usr/sbin/pcs
--Debug Input Start--
{"username": "hacluster", "local": false, "nodes": {"c7-velleros-sde2a": null, "c7-velleros-sde2b": null}, "password": "*****", "force": false}
--Debug Input End--

Finished running: /usr/bin/ruby -I/usr/lib/pcsd/ /usr/lib/pcsd/pcsd-cli.rb auth
Return value: 0
--Debug Stdout Start--
{
  "status": "ok",
  "data": {
    "auth_responses": {
      "c7-velleros-sde2b": {
        "status": "ok",
        "token": "787329eb-4fc8-4618-8e51-4a6d0024619a"
      },
      "c7-velleros-sde2a": {
        "status": "ok",
        "token": "5412dea2-88ab-44ba-a2df-5bbbcd5caffa"
      }
    },
    "sync_successful": true,
    "sync_nodes_err": [
      "mapping",
      "Error",
      "Corosync",
      "Nodes:"
    ],
    "sync_responses": {
      "mapping": {
        "status": "notauthorized"
      },
      "Error": {
        "status": "notauthorized"
      },
      "Corosync": {
        "status": "notauthorized"
      },
      "Nodes:": {
        "status": "notauthorized"
      },
      "c7-velleros-sde2b": {
        "status": "ok",
        "result": {
          "tokens": "accepted"
        }
      }
    }
  },
  "log": [
    "I, [2020-05-14T03:19:28.031623 #3606]  INFO -- : PCSD Debugging enabled\n",
    "D, [2020-05-14T03:19:28.032390 #3606] DEBUG -- : Did not detect RHEL 6\n",
    "D, [2020-05-14T03:19:28.032441 #3606] DEBUG -- : Detected systemd is in use\n",
    "I, [2020-05-14T03:19:28.185975 #3606]  INFO -- : Running: /usr/sbin/corosync-cmapctl totem.cluster_name\n",
    "I, [2020-05-14T03:19:28.186115 #3606]  INFO -- : CIB USER: hacluster, groups: \n",
    "D, [2020-05-14T03:19:28.198682 #3606] DEBUG -- : [\"totem.cluster_name (str) = sdecmsp\\n\"]\n",
    "D, [2020-05-14T03:19:28.198863 #3606] DEBUG -- : []\n",
    "D, [2020-05-14T03:19:28.198925 #3606] DEBUG -- : Duration: 0.012506997s\n",
    "I, [2020-05-14T03:19:28.199063 #3606]  INFO -- : Return Value: 0\n",
    "I, [2020-05-14T03:19:38.464743 #3606]  INFO -- : Running: /usr/sbin/pcs status nodes corosync\n",
    "I, [2020-05-14T03:19:38.464853 #3606]  INFO -- : CIB USER: hacluster, groups: \n",
    "D, [2020-05-14T03:19:38.861563 #3606] DEBUG -- : [\"Error mapping c7-velleros-sde2a\\n\", \"Error mapping c7-velleros-sde2b\\n\", \"Corosync Nodes:\\n\", \" Online:\\n\", \" Offline: c7-velleros-sde2a c7-velleros-sde2b\\n\"]\n",
    "D, [2020-05-14T03:19:38.861746 #3606] DEBUG -- : []\n",
    "D, [2020-05-14T03:19:38.861806 #3606] DEBUG -- : Duration: 0.396678187s\n",
    "I, [2020-05-14T03:19:38.861920 #3606]  INFO -- : Return Value: 0\n",
    "I, [2020-05-14T03:19:38.862554 #3606]  INFO -- : Sending config 'tokens' version 15 254ee28712ecf28799014b4649e5d8ae87984cca to nodes: Error, mapping, c7-velleros-sde2b, Corosync, Nodes:\n",
    "I, [2020-05-14T03:19:38.863785 #3606]  INFO -- : SRWT Node: mapping Request: set_configs\n",
    "E, [2020-05-14T03:19:38.864036 #3606] ERROR -- : Unable to connect to node mapping, no token available\n",
    "I, [2020-05-14T03:19:38.864301 #3606]  INFO -- : SRWT Node: c7-velleros-sde2b Request: set_configs\n",
    "I, [2020-05-14T03:19:38.865980 #3606]  INFO -- : SRWT Node: Error Request: set_configs\n",
    "E, [2020-05-14T03:19:38.866094 #3606] ERROR -- : Unable to connect to node Error, no token available\n",
    "I, [2020-05-14T03:19:38.907007 #3606]  INFO -- : SRWT Node: Corosync Request: set_configs\n",
    "E, [2020-05-14T03:19:38.907138 #3606] ERROR -- : Unable to connect to node Corosync, no token available\n",
    "I, [2020-05-14T03:19:38.907485 #3606]  INFO -- : SRWT Node: Nodes: Request: set_configs\n",
    "E, [2020-05-14T03:19:38.907558 #3606] ERROR -- : Unable to connect to node Nodes:, no token available\n",
    "I, [2020-05-14T03:19:38.949635 #3606]  INFO -- : Sending config response from mapping: {\"status\"=>\"notauthorized\"}\n",
    "I, [2020-05-14T03:19:38.952834 #3606]  INFO -- : Sending config response from Error: {\"status\"=>\"notauthorized\"}\n",
    "I, [2020-05-14T03:19:38.952938 #3606]  INFO -- : Sending config response from Corosync: {\"status\"=>\"notauthorized\"}\n",
    "I, [2020-05-14T03:19:38.953016 #3606]  INFO -- : Sending config response from Nodes:: {\"status\"=>\"notauthorized\"}\n",
    "I, [2020-05-14T03:19:38.953067 #3606]  INFO -- : Sending config response from c7-velleros-sde2b: {\"status\"=>\"ok\", \"result\"=>{\"tokens\"=>\"accepted\"}}\n"
  ]
}

--Debug Stdout End--
--Debug Stderr Start--

--Debug Stderr End--

c7-velleros-sde2a: Authorized
c7-velleros-sde2b: Authorized
Error: Unable to synchronize and save tokens on nodes: mapping, Error, Corosync, Nodes:. Are they authorized?
```

--- 
pcs/corosync/pacemaker makes me crazy :-( btw. it is working on my next servers with os CentOS Linux release 7.6.1810 (Core) and pcs-0.9.165-6.el7.centos.x86_64, corosync-2.4.3-4.el7.x86_64 and pacemaker-1.1.19-8.el7_6.2.x86_64

**Comment by tomjelinek** [comment]
Can you show what this command prints:
`corosync-cmapctl nodelist.node. runtime.totem.pg.mrp.srp.members.`

**Comment by rpolasek** [comment]
**corosync-cmapctl nodelist.node. runtime.totem.pg.mrp.srp.members**
nodelist.node.0.node_id (str) = 1
nodelist.node.0.ring0_addr (str) = c7-velleros-sde2a
nodelist.node.1.node_id (str) = 2
nodelist.node.1.ring0_addr (str) = c7-velleros-sde2b
runtime.totem.pg.mrp.srp.members.3232239394.config_version (u64) = 0
runtime.totem.pg.mrp.srp.members.3232239394.ip (str) = r(0) ip(192.168.15.34) 
runtime.totem.pg.mrp.srp.members.3232239394.join_count (u32) = 1
runtime.totem.pg.mrp.srp.members.3232239394.status (str) = joined
runtime.totem.pg.mrp.srp.members.3232239395.config_version (u64) = 0
runtime.totem.pg.mrp.srp.members.3232239395.ip (str) = r(0) ip(192.168.15.35) 
runtime.totem.pg.mrp.srp.members.3232239395.join_count (u32) = 1
runtime.totem.pg.mrp.srp.members.3232239395.status (str) = joined

**Comment by rpolasek** [comment]
any news, please?

**Comment by tomjelinek** [comment]
The numbers in `runtime.totem.pg.mrp.srp.members` lines don't match `node_id`s for some reason. I asked corosync team about this.

The mismatch causes that the `pcs status nodes corosync` command prints warnings, as can be seen from the log you provided:
```
Error mapping c7-velleros-sde2a
Error mapping c7-velleros-sde2b
Corosync Nodes:
 Online:
 Offline: c7-velleros-sde2a c7-velleros-sde2b
```
There's a bug in pcsd which prevents this to be parsed correctly. Pcsd then considers these to be cluster nodes: mapping, Error, Corosync, Nodes:. Which is obviously wrong.

Before we proceed in fixing this, I would like to hear back from the corosync team.

**Comment by jfriesse** [comment]
@rpolasek Were you using just pcs or also did some hand-editing of corosync.conf? I believe on startup corosync.conf was without nodeids, and then nodeids were added and config was reloaded. Change in node id doesn't affect runtime...srp.members. This is only logical explanation I have for this behavior.

**Comment by rpolasek** [comment]
@jfriesse no i can't use pcs i have prepared a template of my corosync.conf (see above) and during a deployment i change some values like bindnetaddr, node_id and ring0_addr based on the target server (i have to use something like ansible...). everything was working with old crm on centos5. i've had some changes for centos7 but the deployment process is still the same. the problem is that now in centos7 it always failed with the error i've reported and stops the deployment process :-/

**Comment by jfriesse** [comment]
@rpolasek Yup, make sense. You can either decide to use pcs for creating corosync.conf (and it will probably work) or you can deploy corosync.conf from template but then please delete interface subsection (it is no longer needed) to make sure it doesn't interfere with nodelist and make sure to ether not start corosync before final corosync.conf is deployed or restart corosync after corosync.conf is deployed. Because right now it looks like corosync.conf is started without nodelist (or with nodelist where no ring0_addr matches local node ip addr) so corosync generates nodeid based on IP.

**Comment by rpolasek** [comment]
@jfriesse but i'm sure corosync daemon is stopped when i change a corosync.conf and then it is started (after the changes)

**Comment by jfriesse** [comment]
@rpolasek Ok. I think I found it. The main problem is `node_id` in nodelist. It should be `nodeid` (without underscore).

**Comment by rpolasek** [comment]
@jfriesse so it causes all my troubles, right?

**Comment by jfriesse** [comment]
@rpolasek Yes, at least `runtime.totem.pg.mrp.srp.members.` with autogenerated nodeid trouble. I'm really not pcs developer, so not sure what else may break on pcs front.

**Comment by rpolasek** [comment]
@jfriesse and @tomjelinek thanks for your support. i try it try as soon as possible.

**Comment by tomjelinek** [comment]
I took a closer look at this from the pcs side.

The issue has been fixed in pcs-0.10 branch by 7867d17e2ebbc8b5eac7a1b48686461506fc01f7 and 72a2804b355e028b545f859d37ed66c9187d7bd6, meaning all pcs-0.10 releases are unaffected.

Pcs-0.9 branch is in maintenance mode and is only receiving fixes for critical issues. Since the bug described here doesn't manifest itself unless `corosync.conf` has been edited manually in a wrong way, we won't be addressing it there.
